### ISSUE TITLE
Replacing component doesn't trigger node change events

### ISSUE THREAD

**Issue by danielkcz** [BlackDice/scent#17]
The `entity.replace` is the most controversial method which isn't even included in original Ash. Initial idea was to have semantic separation to clearly see in the game code when you actually expect component being replaced. On it's own, it's not that bad idea.

Node types `onAdded` and `onRemoved` methods are not triggered if component of same type is removed and again added during same update cycle. Even methods `remove` && `add` called directly don't trigger that. In general I think it's good approach since nothing has been actually added or removed from entity.

Another point of view is that these change handlers can be used for like one time setup / cleanup of the component within entity. Node types helps to filter out group of components we are interested in and once this group is complete, it's possible to do some setup based on that. This is fine in essence, but component data should be set before being added to entity instead of relying on some node type hidden somewhere.

However there is also case we need to do setup/teardown outside of actual Engine and components, for example to create game objects to be drawn or remove them when entity no longer fulfill node group. This is completely valid point. What I don't agree with is that replacing component (meaning it got different data) should cause another setup. If we need to update game objects based on component data, this should be done exclusively during update loop.

My conclusion here is actually to not change anything as would cause problems and bad use case scenarios of Scent.

**Comment by danielkcz** [comment]
@martikaljuve Can you perhaps add your opinion on this? Do you even use `onAdded` and `onRemoved`. For what situations? And do you replace your components or just update data within them?

**Comment by martikaljuve** [comment]
I knew the `replace` method was there, but I've only needed to update existing components for now. I'm guessing I'd want to have some additional logic if I ever did replace a component, so I'd have a `has` check before actually replacing it.

I use `Node.onAdded` and `Node.onRemoved` in a few places, for example to setup my THREE.js meshes in the RenderSystem. I have cTransform components ('position rotation scale') that actually use THREE.Vector3 and THREE.Euler as the data types, which you probably wouldn't agree with, but it has made it easier to integrate THREE and Scent like this. I have something like this after I have created a THREE.Mesh:

``` js
// in a Node.onAdded callback
var transform = node.transform;
var mesh = new THREE.Mesh(geometry, material);
mesh.position.copy(transform.position); // Copies x, y, z from component Vector3 to mesh Vector3
transform.position = mesh.position; // Replaces component Vector3 with mesh Vector3
```

**Comment by danielkcz** [comment]
I think there is nothing wrong about storing Three objects into components. Yes sure, it would be nice to have just pure data there, but if things like serialization of those components isn't issue for you, then go ahead.

So generally you don't have a issue with `onAdded` and `onRemoved` not being triggered when data changes? I mean if you like remove component and add component of same type within same update cycle, these handlers are not invoked at all. Would you consider that as an issue if you even have a need for that?

**Comment by martikaljuve** [comment]
What other option is there - perhaps to add `Node.onComponentReplaced`? I think I can manage just fine without `onAdded`/`onRemoved` being called after replacing components. In case of some additional setup, I can just add a placeholder component as you mentioned in #2, e.g:

``` js
var cObject3D = new Component('object3d', 'object');
var cSprite = new Component('sprite', 'path color');
var cSpriteChanged = new Component('spriteChanged');

/../

var sprite = new cSprite(['sprite2.png', 0xFFFFFF]);
entity.replace(sprite);
entity.add(cSpriteChanged.pooled());

/../

// in a System
var nChangedSprites = engine.getNodeType(['object3d', 'sprite', 'spriteChanged']);

engine.onUpdate(function() {
    nChangedSprites.each(function(node) {
        node.object3d.object.material = new THREE.SpriteMaterial({
            map: textureLoader.load(node.sprite.path),
            color: node.sprite.color
        });

        node.spriteChanged.dispose();
    });
}
```

**Comment by danielkcz** [comment]
Yeah, this is surely the way, but not that good if you need to do it in multiple places for difference scenarios. Adding `Node.onComponentReplaced` could be one way, but I feel it would make things much more complicated. You see there is this constrain when you are actually updating data (or replacing components) and only after that a node types are updated based on actual state of entities. It would require to track this replacement.

I am thinking about some other approach, basically to keep previous components there and during update loop it would be possible to compare if that component has changed and what is actually different. On top of that if I use some immutable structures that actual share unchanged data, it shouldn't be that problematic in memory department. Just an idea for now.

Either way thanks for your feedback.
